### PR TITLE
[v2] Add support for checkboxes and multiple select

### PR DESCRIPTION
**Pull request by jaredpalmer**
This brings `useField` / `<Field>` / `getFieldProps` inline with similar checkbox and select behavior found in [Vue](https://vuejs.org/v2/guide/forms.html), [Svelte](https://svelte.dev/docs#Element_bindings), and even [Angular](https://netbasal.com/handling-multiple-checkboxes-in-angular-forms-57eb8e846d21)

### Checkboxes
```jsx
// Single checkbox, boolean value.
<Field type="checkbox" name="newsletter" /> // => true/false

// Multiple checkboxes, bound to the same Array.
<Field type="checkbox" name="colors" value="red" /> 
<Field type="checkbox" name="colors" value="green" /> 
<Field type="checkbox" name="colors" value="blue" />  
// => so if they are all checked, will be stored as 
// { colors: ["red", "green", "blue"] } 
// checking/unchecking will add/remove item from the array
```

### Select

```jsx
// Single select 
<Field name="car" as="select">
  <option>Please select one</option>
  <option value="saab">Saab</option>
  <option value="opel">Opel</option>
  <option value="audi">Audi</option>
</Field>

// Multiple select (bound to an Array)
<Field name="car" as="select" multiple>  
  <option value="saab">Saab</option>
  <option value="opel">Opel</option>
  <option value="audi">Audi</option>
</Field> // => { car: ['saab', 'opel', 'audi'] }
```

## Implementation

- Refactor `getFieldProps` to accept an object.
- Allow `useField` to take an object (of your props)
- Do magical stuff when `type="checkbox"`
- Do magical stuff when there is a prop called `multiple`

## Discussion

This is backwards compat, but we may or may not want to call out in the docs that `getFieldProps` includes some magic (if you want it to). 

Closes #527 #211 #360 #1115 

Related to #1525

### CODE REVIEW

File: src/Formik.tsx [hunk 1]
```diff
@@ -921,3 +942,46 @@ function arrayMerge(target: any[], source: any[], options: any): any[] {
...
+
+/** Return the next value for a checkbox */
+function getValueForCheckbox(
+  currentValue: string | any[],
+  checked: boolean,
+  valueProp: any
+) {
+  if (valueProp == 'true' || valueProp == 'false') {
+    return !!checked;
+  } else {
```
Comment (igor0ser) [inline src/Formik.tsx]: `else` is redundant if you have `return`

File: src/Formik.tsx [hunk 2]
```diff
@@ -921,3 +942,46 @@ function arrayMerge(target: any[], source: any[], options: any): any[] {
...
+    if (checked) {
+      return Array.isArray(currentValue)
+        ? currentValue.concat(valueProp)
+        : [valueProp];
+    } else {
+      if (!Array.isArray(currentValue)) {
+        return !!currentValue;
+      }
+      const index = currentValue.indexOf(valueProp);
```
Comment (igor0ser) [inline src/Formik.tsx]: you can use `includes` to instead of `indexOf`

File: src/Formik.tsx [hunk 3]
```diff
@@ -921,3 +942,46 @@ function arrayMerge(target: any[], source: any[], options: any): any[] {
...
+  if (options) {
+    for (let index = 0; index < options.length; index++) {
+      const option = options[index];
+      if (option.selected) {
+        result.push(option.value);
+      }
+    }
+  }
+  return result;
+}
```
Comment (igor0ser) [inline src/Formik.tsx]:
```js
function getSelectedValues(options: any[]) {
   return options.filter(option => option.selected);
 }
```

**Comment by Andreyco** [comment]
@jaredpalmer planning to do a review on this

File: src/Formik.tsx (hunk 3 above)
Comment (jaredpalmer) [reply src/Formik.tsx]: Yep.

File: src/Formik.tsx (hunk 2 above)
Comment (jaredpalmer) [reply src/Formik.tsx]: I need to reuse the index.

**Comment by jaredpalmer** [comment]
I think this all makes a lot of sense, my only one concern is that this could get implemented in userland as `useCheckbox` and `useCheckboxGroup` etc. idk.

File: src/Field.tsx [hunk 4]
```diff
@@ -64,17 +64,33 @@ export interface FieldConfig {
...
   if (process.env.NODE_ENV !== 'production') {
     invariant(
       formik,
       'useField() / <Field /> must be used underneath a <Formik> component or withFormik() higher order component'
     );
   }
-  return formik.getFieldProps<Val>(name, type);
+
+  if (isObject(propsOrFieldName)) {
+    if (process.env.NODE_ENV !== 'production') {
```
Comment (Andreyco) [inline src/Field.tsx]: minor, but I like more __DEV__ constant provided by tsdx (and I know you know)

File: src/Formik.tsx [hunk 5]
```diff
@@ -522,8 +525,10 @@ export function useFormik<Values = object>({
           }
           val = /number|range/.test(type)
             ? ((parsed = parseFloat(value)), isNaN(parsed) ? '' : parsed)
-            : /checkbox/.test(type)
-            ? checked
+            : /checkbox/.test(type) // checkboxes
+            ? getValueForCheckbox(getIn(state.values, field!), checked, value)
+            : !!multiple // <select multiple>
+            ? getSelectedValues(options)
```
Comment (Andreyco) [inline src/Formik.tsx]: too many ternary operators here

**Comment by Andreyco** [comment]
While this is very often requested change, I think it should not be part of formik core.

This change introduces unnecessary noise and possible error surface area into core, while it can be implemented thru user land code, or provided as special set of official extensions (custom fields, hooks, etc).

Exactly as within your implementation, checkboxes represent collections/arrays of selected options, in my opinion. Then, there are special use cases when checkbox serves as a boolean switch. This is exact reason why so many if/else conditions landed in this PR and this could not be generalised in a "proper" way.

How about providing official components that make use of already lean Formik's core?

File: src/Field.tsx (hunk 4 above)
Comment (jaredpalmer) [reply src/Field.tsx]: Let's make a different issue for that.

Comment (Andreyco) [reply src/Field.tsx]: Sure, I can do a find and replace on whole repo

**Comment by jaredpalmer** [comment]
You raise good points and I agree this adds a lot of complexity. However, it has some precedent in the sense that I quite literally implemented the Vue's API.

@eonwhite What do you think? `<Field>` has been magical since day 1.

**Comment by jaredpalmer** [comment]
also ping @nathanforce @crosscompile.

File: src/Field.tsx [hunk 6]
```diff
@@ -64,17 +64,32 @@ export interface FieldConfig {
...
+      invariant(
+        (propsOrFieldName as FieldAttributes<Val>).name,
```
Comment (johnrom) [inline src/Field.tsx]: `isObject` should be defined to narrow the type automatically, like:

```
type NonObject = string | number | boolean | symbol | null | undefined;
const isObject = <T extends object>(maybeObject: NonObject | T): maybeObject is T => 
    typeof value === 'object'; // or whatever isObject's test really is, maybe check for array?
// then
if (isObject(propsOrFieldName)) { // propsOrFieldName is inferred as FieldAttributes<Val> }
```

Comment (jaredpalmer) [reply src/Field.tsx]: Can you open a new issue for this?

**Comment by jaredpalmer** [comment]
I would like to get more eyes on this before going forward. cc @elado @slightlytyler @prichodko

File: src/Field.tsx (hunk 6 above)
Comment (johnrom) [reply src/Field.tsx]: Done, I'll resolve this conversation

#1599

File: src/Formik.tsx [hunk 7]
```diff
@@ -718,20 +718,36 @@ export function useFormik<Values extends FormikValues = FormikValues>({
...
-    ): [FieldInputProps<any>, FieldMetaProps<any>] => {
-      const field = {
+    ({
+      name,
+      type,
+      value: _value,
+      as: is,
+      multiple,
+    }): [FieldInputProps<any>, FieldMetaProps<any>] => {
+      const value = getIn(state.values, name);
```
Comment (crosscompile) [inline src/Formik.tsx]: For this `value` and `_value`, is there a name that more explicitly describes one?  Reading through it's hard to tell what the difference between each is.

Comment (jaredpalmer) [reply src/Formik.tsx]: value vs. valueProp

Comment (crosscompile) [reply src/Formik.tsx]: Ah got it, I think something like `valueState` and `valueProp` would help me differentiate.

Comment (jaredpalmer) [reply src/Formik.tsx]: Yep

**Comment by adrianwix** [comment]
Hi, just a question. I am using a Field type="checkbox" component with a CustomComponent. In Formik I am passing the initialValue of that component like `{ open: false }` but inside the custom component I am getting a PropTypes validation error saying that field.value is undefined. Which would be the expected behavior?
``````

**Comment by jaredpalmer** [comment]
@adrianwix can you create a codesandbox demo so i can take a look?

**Comment by adrianwix** [comment]
@jaredpalmer I just leave my laptop with the project repairing. I can upload it to GitHub and send you the link to the project next Tuesday. The project is not so big I have only implement one page in the client site a not completely and the server but can be completely ignored.

**Comment by adrianwix** [comment]
> @adrianwix can you create a codesandbox demo so i can take a look?

https://codesandbox.io/s/create-react-app-6f7wu That's the example

**Comment by kevinmiguet** [comment]
Somehow I can't make it work with useFormik method `const Home: NextPage<FormikProps<HomeProps>> = (props) => {
  const { values, setFieldValue } = props
  useEffect(() => {
      console.log(values)
    }, [values])
    return (
      <Form>
        Colors
        <Field type="checkbox" name="colors" value="red" /> 
        <Field type="checkbox" name="colors" value="green" /> 
        <Field type="checkbox" name="colors" value="blue" />  
      </Form>
    )
};

const HomeForm = withFormik<Operation[], any>({

  handleSubmit: (values) => {
  },

  displayName: 'HomeFiltersForm',

})(Home);

export default HomeForm;`

and when tickling these checkboxes it just show true or false... am I doing something wrong ?

**Comment by cyonder** [comment]
Why type property isn't passed inside field? Both name and value is passed though. Why do I have to write the type again?

```
<Field name="schedule" type="radio" value="Wow">
    {({ field }) => <Form.Check { ...field } custom type="radio" id="anId" label="Seriously?"/>}
</Field>
```

**Comment by johnrom** [comment]
@cyonder might be better to open a new issue with your question, as this PR is resolved.

**Pull request merged**